### PR TITLE
Fault Handling Pattern 1

### PR DESCRIPTION
**Pull request by scheidere**
This plan (FaultHandlingPattern1.plp) repeatedly executes a 15 degree pan action unless an antenna fault is injected. If an antenna fault is injected, the action repetition stops and the plan waits for it to be resolved indefinitely. If you remove the fault, the original action will resume repetition. 

This plan also calls an initialization node, InitializeLander.plp, which resets pan and tilt to 0 and stows the arm. This should allow for the plan to be run multiple times without restarting the simulator. Additionally, there is logic in the Image node to prevent pan degree boundary (> 360) errors which occur when you let the pan action occur for long enough. 

To test fault handling, observe that panning is occurring, and inject an antenna fault. You should see the pan stop, and it should remain halted until you remove the fault. When you remove it, panning should resume.

To test panning boundary issue, allow the antenna to rotate past 360 degrees. It should not stop.

### CODE REVIEW

File: ow_plexil/src/plans/FaultHandlingPattern1.plp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    Repeat true;
+    Start !Lookup(AntennaFault) && !Initialize;
+
+    log_info ("Running Pattern1Image...");
+    NewAngle = Lookup(PanDegrees) + 15;
+
+    if NewAngle > 360
+    {
+      NumPreviousRevolutions = NewAngle/360;
+      NewAngle = (NumPreviousRevolutions - round(NumPreviousRevolutions))*360;
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern1.plp:32]: PLEXIL has a 'mod' operator, which would be simpler and avoid the rev count variable.

File: ow_plexil/src/plans/FaultHandlingPattern1.plp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  Real NumPreviousRevolutions = 0;
+
+  LibraryCall InitializeLander (initialize = Initialize);
+
+  Pattern1Image:
+  {
+    Repeat true;
+    Start !Lookup(AntennaFault) && !Initialize;
+
+    log_info ("Running Pattern1Image...");
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern1.plp:26]: Suggest omitting this to declutter output.

File: ow_plexil/src/plans/FaultHandlingPattern1.plp [hunk 3]
```diff
@@ -0,0 +1,41 @@
+// The Notices and Disclaimers for Ocean Worlds Autonomy Testbed for Exploration
+// Research and Simulation can be found in README.md in the root directory of
+// this repository.
+
+// Pan in 15 degree increments as long as no antenna fault is present
+// If an antenna fault is present, wait for resolution
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern1.plp:6]: Minor: End sentences with periods.

File: ow_plexil/src/plans/FaultHandlingPattern1.plp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+// No timeout implemented
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern1.plp:7]: I'm not sure what this means.  Either explain or remove.

File: ow_plexil/src/plans/FaultHandlingPattern1.plp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  Boolean Initialize = true;
+  Real NewAngle = 0;
+  Real NumPreviousRevolutions = 0;
+
+  LibraryCall InitializeLander (initialize = Initialize);
+
+  Pattern1Image:
+  {
+    Repeat true;
+    Start !Lookup(AntennaFault) && !Initialize;
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern1.plp:24]: The Initialize variable seems unneeded.  Initialization can be assumed by completion of InitializeLander, which by definition of a CheckedSequence (which is what this is) would have successfully completed before this node.

File: ow_plexil/src/plans/InitializeLander.plp [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+// Research and Simulation can be found in README.md in the root directory of
+// this repository.
+
+// Reset pan and tilt angles to 0 and stow the arm (if not already there)
+
+#include "plan-interface.h"
+
+InitializeLander:
+{
+  InOut Boolean initialize;
```
Comment (kmdalal) [inline ow_plexil/src/plans/InitializeLander.plp:11]: See my comments in calling plan about this variable.  This plan is simple and safe enough to context free and idempotent.

File: ow_plexil/src/plans/InitializeLander.plp [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+  log_info ("Initializing lander...");
+
+  log_info ("Resetting pan angle to 0...");
+  SynchronousCommand pan_antenna (0);
+  log_info ("Resetting tilt angle to 0...");
+  SynchronousCommand tilt_antenna (0);
```
Comment (kmdalal) [inline ow_plexil/src/plans/InitializeLander.plp:18]: Minor point, and not intuitive, but it's better to do the tilt before the pan.  This is because pan can be potentially obstructed by the lander body, but tilt to 0 is always unrestricted.

File: ow_plexil/src/plans/InitializeLander.plp [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  log_info ("If arm unstowed, restowing the arm...");
```
Comment (kmdalal) [inline ow_plexil/src/plans/InitializeLander.plp:19]: Stow is idempotent, so it doesn't matter if the arm was unstowed or not.   You can just say "Stowing the arm".

File: ow_plexil/src/plans/InitializeLander.plp [hunk 9]
```diff
@@ -0,0 +1,25 @@
+// The Notices and Disclaimers for Ocean Worlds Autonomy Testbed for Exploration
+// Research and Simulation can be found in README.md in the root directory of
+// this repository.
+
+// Reset pan and tilt angles to 0 and stow the arm (if not already there)
+
+#include "plan-interface.h"
+
+InitializeLander:
```
Comment (kmdalal) [inline ow_plexil/src/plans/InitializeLander.plp:9]: I think the plan name is too general, and may initially confuse a user.   I'll suggest "InitializeAntennaAndArm".

File: ow_plexil/src/plans/FaultHandlingPattern1.plp (hunk 5 above)
Comment (scheidere) [reply ow_plexil/src/plans/FaultHandlingPattern1.plp:24]: I added this after initial testing because InitializeLander and Pattern1Image are under a concurrence node, and without it panning actions attempt to start concurrently so one gets skipped. Am I missing something else?

File: ow_plexil/src/plans/InitializeLander.plp (hunk 6 above)
Comment (scheidere) [reply ow_plexil/src/plans/InitializeLander.plp:11]: I wanted to keep this initialization node the same for all the plans, which is why this initialize variable is included. It solves an issue when the initialization node is called at the same level as the mission node under a concurrence parent so that the mission doesn't start during initialization. Do you want a different initialization node for this first plan or can I keep this initialize variable so it can stay the same?

File: ow_plexil/src/plans/FaultHandlingPattern1.plp [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    Repeat true;
+    Start !Lookup(AntennaFault);
+
+    NewAngle = Lookup(PanDegrees) + 15;
+
+    if NewAngle > 360
+    {
+      NewAngle = NewAngle mod 360;
+    }
+    endif;
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern1.plp:32]: I think this can simply be `NewAngle = (Lookup(panDegrees) +15) mod 360;`

File: ow_plexil/src/plans/InitializeAntennaAndArm.plp [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  SynchronousCommand tilt_antenna (0);
+  log_info ("Resetting pan angle to 0...");
+  SynchronousCommand pan_antenna (0);
+  log_info ("Stowing the arm...");
+  SynchronousCommand stow();
+
+  log_info ("Initialization complete.");
+  initialize = false;
+
+  Wait 5; // To buffer before a plan begins after initialization
```
Comment (kmdalal) [inline ow_plexil/src/plans/InitializeAntennaAndArm.plp:25]: Why is this needed?

File: ow_plexil/src/plans/InitializeAntennaAndArm.plp [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+// Research and Simulation can be found in README.md in the root directory of
+// this repository.
+
+// Reset pan and tilt angles to 0 and stow the arm (if not already there)
+
+#include "plan-interface.h"
+
+InitializeAntennaAndArm:
+{
+  InOut Boolean initialize;
```
Comment (kmdalal) [inline ow_plexil/src/plans/InitializeAntennaAndArm.plp:11]: I really don't think this variable should be here, as it complicates this plan and the calling plans.   You mentioned concurrence support as the reason, but a cleaner way to that is factor the calls to this node out of the Concurrence.

**Comment by kmdalal** [comment]
The plan ran fine for me.  I've just left code comments that will apply to your other PRs.  We should aim to merge this one first, and then you can merge melodic-devel into your other branches to reduce their code changes.

File: ow_plexil/src/plans/InitializeAntennaAndArm.plp (hunk 11 above)
Comment (scheidere) [reply ow_plexil/src/plans/InitializeAntennaAndArm.plp:25]: After running the plans a bunch sometimes I felt it wasn't visually clear when antenna actions were due to initialization vs. due to the start of the plan itself. Adding a wait gives the user time to process the printed output and visually see that the lander is in the zero positions before it begins moving to execute the plan. I know what these plans will do, but if someone is using this initialization node and writing a new plan it might be easier to debug having a separation like this. That was my logic, but I would be happy to remove it if you think it is not helpful.

File: ow_plexil/src/plans/InitializeAntennaAndArm.plp (hunk 12 above)
Comment (scheidere) [reply ow_plexil/src/plans/InitializeAntennaAndArm.plp:11]: Will that work when the concurrence node is the root node? Currently, all the patterns/plans more complex than Pattern 1 have a concurrence node as the root.

File: ow_plexil/src/plans/InitializeAntennaAndArm.plp (hunk 11 above)
Comment (kmdalal) [reply ow_plexil/src/plans/InitializeAntennaAndArm.plp:25]: The plan does print "Initialization complete", which should delineate this with the test part.  You could reword it to something like "Arm and antenna initialized.".   Waits are best avoided unless they are programatically needed, e.g. giving ROS time to respond or publish a message.

Comment (scheidere) [reply ow_plexil/src/plans/InitializeAntennaAndArm.plp]: Makes sense, I removed the wait.

**Review by kmdalal — APPROVED** [review]
Nice work!
